apps/system/lockscreen/js/widgets/clock/lockscreen_clock_widget.js

```diff
@@ -50,6 +52,44 @@
     this.logger.debug('Clock updated', now);
   };
 
+	LockScreenClockWidget.prototype.updateAlarm =
+	function() {
+    var self = this;
+		// returns alarm info from DataStore
+		navigator.getDataStores('alarms')
+			.then( function(stores){
+        stores[0].getLength().then(function(len){
```

Comment:
Could you just clear up the nested structure with some named functions registered in the widget? Also, for the multiple levels of `if...else`, I feel each branch deserves one method.

It's also necessary to split the data fetching functions out of rendering logic. A fetching method returns Promise should already provide enough interface to chaining the following rendering actions after data fetching in an elegant way, rather than just put a large callback in a `Promise.then` like the old style. For example:

    LockScreenClockWidget.prototype.fetchAlarmData = function() {
        .... return the promise of fetching data, although it's only a one liner.
    }

    LockScreenClockWidget.prototype.renderAlarm = function(fetchingResult) {
      if (fetchingResult.data /*...... to put the conditions here*/) {
        renderNoAlarm();   // there is no alarm according to your code
      } else {
        renderAlarmDateTime();
      }
    }

    // in the using section
   widget.fetchAlarmData()
   .then(widget.renderAlarm)